### DISCUSSION TITLE
Interactive search / filter

### DISCUSSION THREAD

**Discussion by thecsw** [neomutt/neomutt#4140] [Development]
It would be really cool if we can filter messages as we type the search pattern in the search bar.

Right now, when we need to find something in the mailbox with a search pattern, we have to type in the pattern, press enter, and then iterate over the results one-by-one. It would be very nice if we could do it as an active filtering method.

For example, [nnn](https://github.com/jarun/nnn) does the filtering in that way, as when you start typing, it filters out everything that doesn't match, leaving us with only the results. 

The intention is that it would be more intuitive to use, and if I have a mailbox of thousands of emails, I would like to see all results as one whole list. 

I do understand that this change might be challenging in its implementation and it might require some changes in tests, however, I do believe it will make the experience of searching more satisfying and interactive.

**Comment by flatcap** [comment]
Yeah, that's a reasonable idea.
So rather than the `<limit>` command, we'd have a _live_ `<limit-interactive>`.

So, how would you like it to work?

- Would the search be a _regex_?, a _pattern_? _fuzzy_?
  - Would the search use `$simple_search`?
- Which fields of the index would it match against?
  - Would it only match _visible_ fields?

**Comment by ornanovitch** [comment]
Great idea :+1: 

Personal opinion : I love the [FZF](https://github.com/junegunn/fzf) workflow. I'd love having a possibility to filter my mailbox this way.

>     * Would it only match _visible_ fields?

Hard to choose, but for me it's difficult to imagine a situation where I would search in fields that I don't use/see, so I think I would expect this kind of feature to search in visible fields, especially author/subject

**Comment by thecsw** [comment]
@flatcap @Ornanovitch fzf functionality would be superb!

**Comment by flatcap** [comment]
> fzf functionality would be superb

As I understand it, fzf is quite simple.  It's just matching a fuzzy string against fixed strings.
Our situation is more complicated.

The reason I asked about visible fields is that, if a field contains "John Doe", but is truncated to "John", the user will expect "Doe" to match.

A trickier example is `%F` - Author name, or recipient name if the message is from you
Should that match the author's name, even though it's not visible.

> fzf functionality would be superb

So **how** would this work in NeoMutt?
I don't need any technical detail, but I'd like a sample workflow.

Create some sample email data and decide what should happen when the user starts typing.

Have a look through the [index_format docs](https://neomutt.org/guide/reference#index-format) and try to think of any problems.

**Comment by ornanovitch** [comment]
Thanks for the precisions ! I'll try to give some precise ideas ASAP

**Comment by gahr** [comment]
> As I understand it, fzf is quite simple. It's just matching a fuzzy string against fixed strings.
Our situation is more complicated.

As in: we don't have the fixed list of strings in advance, we actually have to *search*, possibly at server side for IMAP.
